src/attention/utils/helpers.ts

```diff
@@ -1,98 +1,243 @@
-import { computePosition, flip, offset, shift, arrow, ReferenceElement } from "@floating-ui/dom";
+import {
+  computePosition,
+  flip,
+  offset,
+  shift,
+  arrow,
+  autoUpdate,
+  ReferenceElement,
+} from '@floating-ui/dom'
 
-const TOP = "top";
-const BOTTOM = "bottom";
-const LEFT = "left";
-const RIGHT = "right";
-type Directions = "top" | "right" | "bottom" | "left";
+const TOPSTART = 'top-start'
+const TOP = 'top'
+const TOPEND = 'top-end'
+const RIGHTSTART = 'right-start'
+const RIGHT = 'right'
+const RIGHTEND = 'right-end'
+const BOTTOMSTART = 'bottom-start'
+const BOTTOM = 'bottom'
+const BOTTOMEND = 'bottom-end'
+const LEFTSTART = 'left-start'
+const LEFT = 'left'
+const LEFTEND = 'left-end'
+
+type Directions =
+  | 'top'
+  | 'top-start'
+  | 'top-end'
+  | 'right'
+  | 'right-start'
+  | 'right-end'
+  | 'bottom'
+  | 'bottom-start'
+  | 'bottom-end'
+  | 'left'
+  | 'left-start'
+  | 'left-end'
 
 export const opposites = {
+  [TOPSTART]: BOTTOMEND,
   [TOP]: BOTTOM,
+  [TOPEND]: BOTTOMSTART,
+  [BOTTOMSTART]: TOPEND,
   [BOTTOM]: TOP,
+  [BOTTOMEND]: TOPSTART,
+  [LEFTSTART]: RIGHTEND,
   [LEFT]: RIGHT,
+  [LEFTEND]: RIGHTSTART,
+  [RIGHTSTART]: LEFTEND,
   [RIGHT]: LEFT,
-};
+  [RIGHTEND]: LEFTSTART,
+}
+
 export const arrowLabels = {
-  [TOP]: "↑",
-  [BOTTOM]: "↓",
-  [LEFT]: "←",
-  [RIGHT]: "→",
-};
-export const directions = [TOP, BOTTOM, LEFT, RIGHT];
-export const rotation = {
+  [TOPSTART]: '↑',
+  [TOP]: '↑',
+  [TOPEND]: '↑',
+  [BOTTOMSTART]: '↓',
+  [BOTTOM]: '↓',
+  [BOTTOMEND]: '↓',
+  [LEFTSTART]: '←',
+  [LEFT]: '←',
+  [LEFTEND]: '←',
+  [RIGHTSTART]: '→',
+  [RIGHT]: '→',
+  [RIGHTEND]: '→',
+}
+export const directions = [
+  TOPSTART,
+  TOP,
+  TOPEND,
+  BOTTOMSTART,
+  BOTTOM,
+  BOTTOMEND,
+  LEFTSTART,
+  LEFT,
+  LEFTEND,
+  RIGHTSTART,
+  RIGHT,
+  RIGHTEND,
+]
+export const rotation: any = {
```

Comment:
was a type required here? `any` doesn't seem to help much 😄

Reply:
I added type `any` to rotation because it was complaining on line 130, when I'm setting `arrowRotation` to `rotation[arrowDirection]`: 
```
"Element implicitly has an 'any' type because expression of type 'string' can't be used to index type '{ "left-start": number; left: number; "left-end": number; "top-start": number; top: number; "top-end": number; "right-start": number; right: number; "right-end": number; "bottom-start": number; bottom: number; "bottom-end": number; }'.
  No index signature with a parameter of type 'string' was found on type '{ "left-start": number; left: number; "left-end": number; "top-start": number; top: number; "top-end": number; "right-start": number; right: number; "right-end": number; "bottom-start": number; bottom: number; "bottom-end": number; }'."
```

Reply:
Maybe something like `Record<Directions, number>`?

Reply:
I tried that! But then I also need to set the direction enums to of type `Directions`: 
```
const TOP_START: Directions = 'top-start'
const TOP: Directions = 'top'
const TOP_END: Directions = 'top-end'
const RIGHT_START: Directions = 'right-start'
const RIGHT: Directions = 'right'
const RIGHT_END: Directions = 'right-end'
const BOTTOM_START: Directions = 'bottom-start'
const BOTTOM: Directions = 'bottom'
const BOTTOM_END: Directions = 'bottom-end'
const LEFT_START: Directions = 'left-start'
const LEFT: Directions = 'left'
const LEFT_END: Directions = 'left-end'
```
That seems to remove the error for when I'm setting arrowRotation to rotation[arrowDirection]. However, now I instead get an error regarding the type of the `name`-parameter that is being passed into `isDirectionVertical`:
```
const isDirectionVertical = (name: Directions) =>
  [TOP_START, TOP, TOP_END, BOTTOM_START, BOTTOM, BOTTOM_END].includes(name)
``` 
I get this error then: 
```
Argument of type 'Directions' is not assignable to parameter of type '"top" | "top-start" | "top-end" | "bottom" | "bottom-start" | "bottom-end"'.
  Type '"right"' is not assignable to type '"top" | "top-start" | "top-end" | "bottom" | "bottom-start" | "bottom-end"'.
```

Reply:
Updated in this commit: [2a7276c](https://github.com/warp-ds/core/pull/15/commits/2a7276c3784d4a7cb8e14b7a1a790a167781024a#diff-f7f359dac756e644d43cfb407c55dca65dc06f8c392dd99f6a13681375c1cc10R115)